rbx_types/src/unique_id.rs

```diff
@@ -0,0 +1,211 @@
+    random: i64,
+}
+
+static INDEX: AtomicU32 = AtomicU32::new(0);
+
+impl UniqueId {
+    pub fn new(index: u32, time: u32, random: i64) -> Self {
+        Self {
+            index,
+            time,
+            random,
+        }
+    }
+
+    pub fn now() -> Result<Self, UniqueIdError> {
+        let time = SystemTime::now()
+            .duration_since(*EPOCH)
+            .map_err(|_| UniqueIdError::SystemPastTime)?;
+
+        Ok(Self {
+            index: INDEX.fetch_add(1, Ordering::AcqRel),
+            time: u32::try_from(time.as_secs()).map_err(|_| UniqueIdError::Overflow)?,
+            // This matches Roblox's behavior, where the value is both an i64
+            // but is also always positive.
+            random: thread_rng().gen_range(0..i64::MAX),
+        })
+    }
+
+    pub fn time(&self) -> u32 {
+        self.time
+    }
+
+    pub fn index(&self) -> u32 {
+        self.index
+    }
+
+    pub fn random(&self) -> i64 {
+        self.random
+    }
+}
+
+impl fmt::Display for UniqueId {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        write!(f, "{:016x}{:08x}{:08x}", self.random, self.time, self.index)
+    }
+}
+
+#[cfg(feature = "serde")]
+mod serde_impl {
+    use super::UniqueId;
+    use serde::{de, Deserialize, Deserializer, Serialize, Serializer};
+    use std::{convert::TryInto, fmt};
+
+    impl Serialize for UniqueId {
+        fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
+        where
+            S: Serializer,
+        {
+            if serializer.is_human_readable() {
+                serializer.serialize_str(&self.to_string())
+            } else {
+                let mut bytes = Vec::with_capacity(16);
+                bytes.extend_from_slice(&self.random.to_be_bytes());
+                bytes.extend_from_slice(&self.time.to_be_bytes());
+                bytes.extend_from_slice(&self.index.to_be_bytes());
+
+                serializer.serialize_bytes(&bytes)
+            }
+        }
+    }
+
+    impl<'de> Deserialize<'de> for UniqueId {
+        fn deserialize<D>(deserializer: D) -> Result<Self, D::Error>
+        where
+            D: Deserializer<'de>,
+        {
+            if deserializer.is_human_readable() {
+                deserializer.deserialize_str(HumanVisitor)
+            } else {
+                deserializer.deserialize_bytes(NonHumanVisitor)
+            }
+        }
+    }
+
+    struct HumanVisitor;
+    struct NonHumanVisitor;
+
+    impl<'de> de::Visitor<'de> for HumanVisitor {
+        type Value = UniqueId;
+
+        fn expecting(&self, formatter: &mut fmt::Formatter) -> fmt::Result {
+            write!(formatter, "a sequence of 32 hexadecimal characters")
+        }
+
+        fn visit_str<E>(self, v: &str) -> Result<Self::Value, E>
+        where
+            E: de::Error,
+        {
+            if v.len() == 32 {
+                Ok(UniqueId {
+                    random: i64::from_str_radix(&v[0..16], 16).map_err(E::custom)?,
```

Comment:
There's some code duplication between rbx_xml and rbx_types now... maybe we should `impl FromStr for UniqueId`?

Reply:
I'm not thrilled with how this ends up looking but I've done it. The error handling in rbx_xml is bad.